tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/hlo_canonicalize_reduction.cc

```diff
@@ -0,0 +1,238 @@
+        if (!dims_to_reduce_set.count(i)) dims_to_keep.push_back(i);
+      }
+
+      OpBuilder b(op);
+      auto loc = op.getLoc();
+      // TODO(disc): uniformed shape_scalar_type with shape_derivation
+      auto shape_scalar_type = b.getIntegerType(32);
+      auto one = b.create<ConstantIntOp>(loc, 1ll, shape_scalar_type);
+
+      // funtion to get total elements in selected dimensions
+      auto dim_prod = [&](ArrayRef<int64_t> dims) {
+        Value nelems = one;
+        for (int64_t v : dims) {
+          Value dim_index = b.create<memref::DimOp>(loc, op.getOperand(0), v);
+          nelems = b.create<MulIOp>(
+              loc, nelems,
+              b.create<IndexCastOp>(loc, dim_index, shape_scalar_type));
+        }
+        return nelems;
+      };
+
+      SmallVector<Value, 2> new_operand_dims;
+      DenseIntElementsAttr attr;
+      Value nelem_to_reduce = dim_prod(dims_to_reduce);
+      Value nelem_to_keep = dim_prod(dims_to_keep);
+      if (rank == ndims_to_reduce) {
+        // case c) Reduce to scalar.
+        // Currently we don't support reduce to scalar directly.
+        // As a workaround, we convert the `reduce to scalar` to a rank 2
+        // column reduction having following form:
+        // Suppose nelems = ProdutionOp(ShapeOp(I)), We convert I into
+        // shape `[nelems, 1]`.
+        // TODO(disc): this may have performance issue. Implements a reduce to
+        // scalar schedule if necessnary.
+        new_operand_dims.push_back(nelem_to_reduce);
+        new_operand_dims.push_back(nelem_to_keep);
+        attr = DenseIntElementsAttr::get(
+            RankedTensorType::get({1}, b.getIntegerType(64)), {0ll});
+      } else if (dims_to_reduce[0] == 0) {
+        // case a) column reduction
+        new_operand_dims.push_back(nelem_to_reduce);
+        new_operand_dims.push_back(nelem_to_keep);
+        attr = DenseIntElementsAttr::get(
+            RankedTensorType::get({1}, b.getIntegerType(64)), {0ll});
+      } else {
+        // case b) raw reduction
+        new_operand_dims.push_back(nelem_to_keep);
+        new_operand_dims.push_back(nelem_to_reduce);
+        attr = DenseIntElementsAttr::get(
+            RankedTensorType::get({1}, b.getIntegerType(64)), {1ll});
+      }
+
+      Value new_operand_shape =
+          b.create<tensor::FromElementsOp>(loc, new_operand_dims);
+
+      SmallVector<Value, 4> new_operands;
+      for (Value operand : op.inputs()) {
+        new_operands.push_back(b.create<DynamicReshapeOp>(
+            loc,
+            RankedTensorType::get(
+                SmallVector<int64_t, 4>(new_operand_dims.size(),
+                                        ShapedType::kDynamicSize),
+                elem_ty),
+            operand, new_operand_shape));
+      }
+      auto new_op =
+          b.create<ReduceOp>(loc, new_operands, op.init_values(), attr);
+      BlockAndValueMapping mapping;
+      op.body().cloneInto(&new_op.body(), new_op.body().end(), mapping);
```

Comment:
Instead of cloning, we should be able to move the body instead: moving a block from one region to another is cheap. Then is it just looping of the block arguments together with `llvm::zip` and calling `replaceAllUsesWith` on each pair.

Reply:
can I use `new_op.body().takeBody(op.body());` to move the whole body to the new_op? it seems work.